sql/core/src/main/scala/org/apache/spark/sql/api/r/SQLUtils.scala

```diff
@@ -145,16 +145,50 @@ private[sql] object SQLUtils {
       packageNames: Array[Byte],
       broadcastVars: Array[Object],
       schema: StructType): DataFrame = {
-    val bv = broadcastVars.map(x => x.asInstanceOf[Broadcast[Object]])
-    val realSchema =
-      if (schema == null) {
-        SERIALIZED_R_DATA_SCHEMA
-      } else {
-        schema
-      }
+    val bv = broadcastVars.map(_.asInstanceOf[Broadcast[Object]])
+    val realSchema = if (schema == null) SERIALIZED_R_DATA_SCHEMA else schema
     df.mapPartitionsInR(func, packageNames, bv, realSchema)
   }
 
+  /**
+   * The helper function for gapply() on R side.
+   */
+  def gapply(
+      df: DataFrame,
+      func: Array[Byte],
+      packageNames: Array[Byte],
+      broadcastVars: Array[Object],
+      schema: StructType,
+      col: String): DataFrame = {
+    val realSchema = if (schema == null) SERIALIZED_R_DATA_SCHEMA else schema
+    val dfSchema = df.select(df(col)).schema
+    if (dfSchema.length == 0) throw new IllegalArgumentException(s"Invaid column name $col")
+    val dataType = dfSchema(0).dataType
+    val sqlContext = df.sqlContext
+
+    import sqlContext.implicits._
+
```

Comment:
The logic of gapply() could be:
  new a RelationGroupedDataset as rgd
  call rgd.flatMapGroupsInR to create updated logical plan